File: packages/gavio-py/gavio/interceptors/audit/sinks/stdout.py
"""StdoutSink — human-readable audit output for development (F-OBS-05)."""

from __future__ import annotations

import json
import sys

from ..record import AuditRecord
from ..sink import AuditSink


class StdoutSink(AuditSink):
    """Print each audit record to stdout. Zero dependencies."""

    def __init__(self, pretty: bool = True, stream=None) -> None:
        self.pretty = pretty
        self._stream = stream if stream is not None else sys.stdout

    async def write(self, record: AuditRecord) -> None:
        data = record.to_dict()
        if self.pretty:
            line = self._format_pretty(data)
        else:
            line = json.dumps(data)
        print(line, file=self._stream, flush=True)

    @staticmethod
    def _format_pretty(data: dict) -> str:
        usage = data["token_usage"]
        pii = data["pii_entity_types"] or ["none"]
        return (
            "[gavio:audit] "
            f"trace={data['trace_id'][:18]}… "
            f"{data['provider']}/{data['model']} "
            f"tokens={usage['total_tokens']} "
            f"cost=${data['cost_usd']:.6f} "
            f"latency={data['latency_ms']}ms "
            f"cache={'HIT' if data['cache_hit'] else 'miss'} "
            f"pii={','.join(pii)} "
            f"interceptors=[{','.join(data['interceptors_fired'])}]"
        )

File: packages/gavio-js/src/interceptors/audit/sinks/stdout.ts
/** stdoutSink — human-readable audit output for development (F-OBS-05). */

import type { AuditRecord } from '../record.js'
import type { AuditSink } from '../sink.js'

export interface StdoutSinkOptions {
  pretty?: boolean
  write?: (line: string) => void
}

/** Print each audit record to stdout. Zero dependencies. */
export function stdoutSink(options: StdoutSinkOptions = {}): AuditSink {
  const pretty = options.pretty ?? true
  // eslint-disable-next-line no-console
  const emit = options.write ?? ((line: string) => console.log(line))
  return {
    async write(record: AuditRecord): Promise<void> {
      const data = record.toJSON()
      emit(pretty ? formatPretty(data) : JSON.stringify(data))
    },
  }
}

function formatPretty(data: Record<string, unknown>): string {
  const usage = data['tokenUsage'] as { totalTokens: number }
  const piiTypes = data['piiEntityTypes'] as string[]
  const pii = piiTypes.length > 0 ? piiTypes : ['none']
  const interceptors = data['interceptorsFired'] as string[]
  const traceId = String(data['traceId'])
  const cost = Number(data['costUsd'])
  return (
    '[gavio:audit] ' +
    `trace=${traceId.slice(0, 18)}… ` +
    `${String(data['provider'])}/${String(data['model'])} ` +
    `tokens=${usage.totalTokens} ` +
    `cost=$${cost.toFixed(6)} ` +
    `latency=${String(data['latencyMs'])}ms ` +
    `cache=${data['cacheHit'] ? 'HIT' : 'miss'} ` +
    `pii=${pii.join(',')} ` +
    `interceptors=[${interceptors.join(',')}]`
  )
}
